mcfly.zsh

```diff
@@ -0,0 +1,103 @@
+#!/bin/zsh
+
+# Ensure stdin is a tty
+[[ -t 0 ]] || return
+
+# Avoid loading this file more than once
+if [[ "$__MCFLY_LOADED" == "loaded" ]]; then
+  return 0
+fi
+__MCFLY_LOADED="loaded"
+
+# Ensure HISTFILE exists.
+export HISTFILE="${HISTFILE:-$HOME/.zsh_history}"
+if [[ ! -r "${HISTFILE}" ]]; then
+  echo "McFly: ${HISTFILE} does not exist or is not readable. Please fix this or set HISTFILE to something else before using McFly."
+  return 1
+fi
+
+# MCFLY_SESSION_ID is used by McFly internally to keep track of the commands from a particular terminal session.
+export MCFLY_SESSION_ID=$(dd if=/dev/urandom bs=256 count=1 2> /dev/null | env LC_ALL=C tr -dc 'a-zA-Z0-9' | head -c 24)
+
+setopt extended_history       # record timestamp of command in HISTFILE
+setopt hist_expire_dups_first # delete duplicates first when HISTFILE size exceeds HISTSIZE
+setopt hist_ignore_dups       # ignore duplicated commands history list
+setopt hist_ignore_space      # ignore commands that start with space
+setopt inc_append_history     # add commands to HISTFILE in order of execution
+setopt share_history          # share command history data
+setopt interactive_comments   # allow comments in interactive shells (like Bash does)
+
+# Setup a function to clear the shell history.
+function erase_history { local HISTSIZE=0; }
+
+# Setup a function to be used by $PROMPT_COMMAND.
+function mcfly_prompt_command {
+  local exit_code=$? # Record exit status of previous command.
+
+  # Populate McFly's temporary, per-session history file from recent commands in the shell's primary HISTFILE.
+  if [[ ! -f "${MCFLY_HISTORY}" ]]; then
+    export MCFLY_HISTORY=$(mktemp -t mcfly.XXXXXXXX)
+    tail -n100 "${HISTFILE}" >| ${MCFLY_HISTORY}
```

Comment:
It's best to use `fc` than to directly read the user history for the following reasons:
- Some commands may span multiple lines
- `fc` may contain more entries than `HISTFILE`. For example, the user may populate `fc` from multiple history files, each used for a specific machine, and sync them between machines.
So a user may have `.zsh_history_machine1` that holds the commands history typed on machine1 and `.zsh_history_machine2` that holds the commands history typed on machine2. Then he can sync them, and load them on zsh startup so that `fc` contains on each of the machines.
- You will avoid the call to `tail` which is good for performance (though probably negligible)